zabbix_smartmontools/__init__.py

```diff
@@ -1,150 +1,204 @@
+    jsonData = []
+    for (device, protocol) in diskList:
+        if config['mode'] == 'serial' or config['skipDuplicates']:
+            serial = getSerial(device)
         else:
-            startStopResult = 'UNKNOWN'
-
-        sender.append('%s smartctl.value[%s,startStop] "%s"' % (host, dR, startStopResult))
-
-        startStopMaxRe = re.search(r'^Recommended maximum start stop count:\s+(\d+)|^Specified cycle count over device lifetime:\s+(\d+)', p, re.M | re.I)
-        if startStopMaxRe:
-            if startStopMaxRe.group(1):
-                startStopMaxResult = startStopMaxRe.group(1)
-            elif startStopMaxRe.group(2):
-                startStopMaxResult = startStopMaxRe.group(2)
+            serial = None
+        if config['mode'] == 'serial' and serial:
+            identifier = serial
+        else:
+            identifier = device
+        jsonData.append({'{#DDRIVESTATUS}': identifier})
+        if not serial in reported_serials or not config['skipDuplicates']:
+            jsonData.append({"{#DISKID}": identifier})
+            jsonData.append({"{#DISKIDSAS}": identifier})
+        if serial:
+            reported_serials.add(serial)
+    return jsonData
+
+
+# Get the serial number of a hard drive, using the fastest method available on
+# each platform.
+def getSerial(device):
+    """ Get the serial number of a hard drive """
+    serial = None
+    if 'freebsd' in sys.platform:
+        DISK_IDENT_SIZE = 256       # from sys/disk.h
+        DIOCGIDENT = 0x41006489     # from sys/disk.h
+        ident = DISK_IDENT_SIZE * r" "
+        f = open("/dev/%s" % device, 'r')
+        try:
+            serial = ioctl(f, DIOCGIDENT, ident).decode().strip("\x00")
+        except OSError as e:
+            if e.errno == errno.EBADF:
+                # EBADF indicates something that's not a disk, such as a SES
+                # enclosure
+                True
             else:
-                startStopMaxResult = 'UNKNOWN'
-
-            sender.append('%s smartctl.value[%s,startStopMax] "%s"' % (host, dR, startStopMaxResult))
-
-    defectsRe = re.search(r'^Elements in grown defect list:\s+(\d+)', p, re.M | re.I)
-    if defectsRe:
-        sender.append('%s smartctl.value[%s,defects] "%s"' % (host, dR, defectsRe.group(1)))
-
-    poweredHoursRe = re.search(r'^number of hours powered up \=\s+(\d+)', p, re.M | re.I)
-    if poweredHoursRe:
-        sender.append('%s smartctl.value[%s,poweredHours] "%s"' % (host, dR, poweredHoursRe.group(1)))
-
-    nonMediumErrorsRe = re.search(r'^Non-medium error count:\s+(\d+)', p, re.M | re.I)
-    if nonMediumErrorsRe:
-        sender.append('%s smartctl.value[%s,nonMediumErrors] "%s"' % (host, dR, nonMediumErrorsRe.group(1)))
-
-    if sender:
-        error = False
+                raise e
+    elif sys.platform.startswith('linux'):
+        # TODO: is there a way to get this without using a subprocess?
+        cp = subprocess.run(["/sbin/udevadm", "info", "--query=all",
```

Comment:
```suggestion
        try:
            cp = subprocess.check_output(["/sbin/udevadm", "info", "--query=all",
            "--name=/dev/%s" % device])
        except:
            raise
        for line in cp.decode().splitlines():
```
`capture_output` is python37, staying with 33.

I can't recommend the fastest option for Linux. FreeBSD version alone is insane enough for me.

Reply:
Also please address other `run`s.

Reply:
Nope. [subprocess.run ](https://docs.python.org/3.5/library/subprocess.html#subprocess.run)itself is python35. Could you consider [subprocess.check_output](https://docs.python.org/3.3/library/subprocess.html#subprocess.check_output)?